rust/connlib/tunnel/src/client.rs

```
@@ -648,10 +649,28 @@ impl ClientState {
         &mut self,
         query: DnsQuery<'static>,
         response: Result<
-            Result<hickory_resolver::lookup::Lookup, hickory_resolver::error::ResolveError>,
-            futures_bounded::Timeout,
+            Result<
+                Result<hickory_resolver::lookup::Lookup, hickory_resolver::error::ResolveError>,
+                futures_bounded::Timeout,
+            >,
+            DnsQueryError,
         >,
```

Trying to understand this type
`Result <Result <Result <HickoryLookup, HickoryResolveError>, Timeout>, DnsQueryError>`

Formerly `Result <HickoryLookup, HickoryResolveError>, Timeout>`

But these `Result`s are taken as _input_, so this function must be for wrapping them into a packet and immediately sending the packet back toward the user? (I don't want to say "client" since there's at least 3 components that act as a client)

So the type means
- If everything is good we got a lookup, e.g. `example.com --> 1.1.1.1`
- If the upstream returns an error, like NXDOMAIN we got a `ResolveError`
- If the upstream never replied, we have a timeout (Why is that a future and not a `Duration` though?)
- If the query made no sense in the first place and we did not even try to forward it, we have a `DnsQueryError`

It kind of makes sense, but it took a few minutes to read this signature.

---

> Trying to understand this type `Result <Result <Result <HickoryLookup, HickoryResolveError>, Timeout>, DnsQueryError>`

:sweat_smile: 

The the way this came about is that the sans-IO `ClientState` needs to be able to deal with all sorts of errors that happen to the side-effect of making a DNS query.

1. The outer-most result is the error that happened when _scheduling_ a DNS query. That is being added in this PR.
2. The 2nd layer of `Result` represents a timeout on the async operation of resolving the DNS query. We should probably return an error for that to the client as well.
3. The 3rd and inner-most `Result` represents the result of the DNS resolution as the `hickory-dns` library reports it.

Does this make sense?

---

Yeah, that's what I thought, it just took a while to figure out haha